spec/controllers/case_workers/admin/management_information_controller_spec.rb

```diff
@@ -26,65 +23,50 @@
     end
 
     describe '#GET download' do
-      context 'for a valid report type' do
-        let(:report_type) { 'management_information' }
+      subject(:download) { get :download, params: { report_type: report_type } }
 
-        context 'using DB storage' do
-          let(:content) { 'header1,header2,header3' }
-          let!(:stats_report) { create(:stats_report, report_name: 'management_information', report: content) }
-
-          before do
-            get :download, params: { report_type: report_type }
-          end
+      let(:report_type) { 'management_information' }
 
-          it 'returns http success' do
-            expect(response).to be_successful
-          end
+      context 'when the report type is valid' do
+        let(:disk_service_prefix) { 'rails/active_storage/disk' }
+        let(:filename) { 'mi_report.csv' }
 
-          it 'renders the template' do
-            expect(response.headers['Content-Type']).to eq 'text/csv'
+        before do
+          create(:stats_report, report_name: report_type).tap do |report|
+            report.document.attach(io: StringIO.new, filename: filename)
           end
+          download
         end
 
-        context 'using S3 storage' do
-          let(:content) { 'header1,header2,header3' }
-          let(:document) { StringIO.new(content) }
-          let!(:stats_report) {
-            create(
-              :stats_report,
-              report_name: report_type,
-              document: document,
-              document_file_name: "#{report_type}_#{Time.now.to_s(:number)}.csv",
-              document_content_type: 'text/csv'
-            )
-          }
-
-          before do
-            get :download, params: { report_type: report_type }
-          end
+        it 'redirects to the service url of the document' do
```

Comment:
Is it possible to do something more idiomatic like:

```
let(:stats_report) do
  create(:stats_report, report_name: report_type).tap do |report|
    report.document.attach(io: StringIO.new, filename: filename)
  end
end

before { download }
...
expect(response).to redirect_to(stats_report.document.blob.service_url)
```

Reply:
ok, makes sense.

could try 
```
allow(stats_report).to receive_message_chain(:document, :blob, :service_url).and_return('my-expiring-object-url')
```

but perhaps a regex matcher is better after all and perhaps with a separate test to match any expiry limit of on the url. maybe excessive though?! 
```
expect(response.location).to match(%r{expiry=blah})
```

Reply:
I found that `receive_message_chain` gets flagged by Rubocop (see [Cop/RSpec/MessageChain](https://www.rubydoc.info/gems/rubocop-rspec/1.7.0/RuboCop/Cop/RSpec/MessageChain)) and, searching around, it does seem to be discouraged (see [here](https://stackoverflow.com/questions/53648078/rspec-rubocop-why-receive-message-chain-is-a-code-smell)). I have achieved the same with:

```ruby
allow(Stats::StatsReport).to receive(:most_recent_by_type).and_return(stats_report)
allow(stats_report.document.blob).to receive(:service_url).and_return(test_url)
```

I'm not entirely convinced of this but it works.